File: kotlin/src/main/kotlin/Day02.kt
fun main() {
    val input = util.readInput("Day02.txt")

    fun lineToGame(line: String): List<Map<String, Int>> {
        return line.split(";").map {
            it.split(",")
                .map { it.trim() }
                .map {
                    val (n, color) = it.split(" ")
                    color to n.toInt()
                }
                .toMap()
        }
    }

    fun part1(input: List<String>): Int {
        fun checkGame(game: List<Map<String, Int>>): Boolean =
            game
                .filter {
                    val n = it["red"] ?: 0
                    n <= 12
                }
                .filter {
                    val n = it["green"] ?: 0
                    n <= 13
                }
                .filter {
                    val n = it["blue"] ?: 0
                    n <= 14
                }
                .size == game.size

        return input.sumOf {
            val colonPos = it.indexOf(':') 
            val id = it.substring(5, colonPos)
            val rest = it.substring(colonPos + 1)
            val game = lineToGame(rest)
            if (checkGame(game)) id.toInt() else 0
        }
    }

    fun part2(input: List<String>): Int {
        fun minimumCubes(game: List<Map<String, Int>>): Int =
            game.map { it.get("red") ?: 1 }.max() *
                game.map { it.get("green") ?: 1 }.max() *
                game.map { it.get("blue") ?: 1 }.max()

        return input.sumOf {
            val colonPos = it.indexOf(':') 
            val rest = it.substring(colonPos + 1)
            val game = lineToGame(rest)
            minimumCubes(game)
        }
    }

    part1(input).dbg()
    part2(input).dbg()
}

File: scala/src/Day02.scala
import cats.effect.IOApp
import cats.effect.IO
import utils.readInput
import fs2.Stream

object Day02 extends IOApp.Simple:
  val MAX_RED = 12
  val MAX_GREEN = 13
  val MAX_BLUE = 14

  private def parseToGame(str: String): Array[Map[String, Int]] =
    str
      .split(";")
      .map(
        _.split(",")
          .map(_.strip)
          .map(_ match
            case s"$n $color" => color -> n.toInt
          )
          .toMap
          .withDefaultValue(0)
      )

  private def mapLinesToSum(
      lines: Stream[IO, String],
      f: (Int, String) => Int
  ): Stream[IO, Int] =
    lines
      .map(_ match
        case s"Game $id: $rest" => f(id.toInt, rest)
        case _                  => 0
      )
      .foldMonoid
      .head

  private def part1(input: Stream[IO, String]) =
    def idIfPossible(id: Int, game: String): Int =
      val parsed = parseToGame(game)
      if parsed
          .filter(_("red") <= MAX_RED)
          .filter(_("green") <= MAX_GREEN)
          .filter(_("blue") <= MAX_BLUE)
          .length == parsed.length
      then id
      else 0

    mapLinesToSum(input, idIfPossible)

  private def part2(input: Stream[IO, String]) =
    def power(id: Int, game: String): Int =
      val parsed = parseToGame(game)
      parsed.map(_("red")).max *
        parsed.map(_("green")).max *
        parsed.map(_("blue")).max

    mapLinesToSum(input, power)

  def run: IO[Unit] =
    val input = readInput("Day02")

    part1(input)
      .zip(part2(input))
      .evalTap(IO.println)
      .compile
      .drain
